Fixturize tests/frame/test_arithmetic

h-vetinari opened the pull request:
Split off from #22730 as per review from @WillAyd 

The changes in `conftest.py` are due to the following:
> In translating the quasi-fixtures from `TestData` to `conftest` in #22236, I sorted the dtypes for the columns of `mixed_float_frame` and `mixed_int_frame`, which turns out to have been a mistake. This is reverted here to be a true translation of the attribute of `TestData`. Otherwise, tests in the newly fixturized `test_arithmetic.py` would fail.

pep8speaks commented:
Hello @h-vetinari! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/tests/frame/conftest.py`](https://github.com/h-vetinari/pandas/blob/8d5c3074078924c8da7afa06f821ec5de6b0cd4f/pandas/tests/frame/conftest.py) !

 - There are no PEP8 issues in the file [`pandas/tests/frame/test_arithmetic.py`](https://github.com/h-vetinari/pandas/blob/8d5c3074078924c8da7afa06f821ec5de6b0cd4f/pandas/tests/frame/test_arithmetic.py) !

h-vetinari committed:
Fixturize tests/frame/test_arithmetic

h-vetinari committed:
PEP8 fix

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,97 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-    def test_arith_flex_frame(self):
-        seriesd = tm.getSeriesData()
-        frame = pd.DataFrame(seriesd).copy()
+    @pytest.mark.parametrize('op', ['add', 'sub', 'mul', 'div', 'truediv',
```
can you change these to use the fixture: all_arithmetic_operators

h-vetinari replied:
will do

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,97 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-                                dtype=np.int64)
-
-        ops = ['add', 'sub', 'mul', 'div', 'truediv', 'pow', 'floordiv', 'mod']
         if not PY3:
```
you prob don't need this (when using the fixture)

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,97 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+        tm.assert_frame_equal(result, exp)
+        _check_mixed_float(result, dtype=dict(C=None))
+
+        # vs mix int
```
might be more clear to move these to a separate test (from here down)

h-vetinari replied:
I don't want to get into orthogonal changes here. I haven't changed anything about this except removing the indentation. This goes for the test split requested below as well.

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,97 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-
-        const_add = frame.add(1)
-        tm.assert_frame_equal(const_add, frame + 1)
+        # ndim >= 3
```
here and down could be in the top-section, e.g. the first test as they test all ops (or another tests is even better, with the fixture this becomes easy)

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,97 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-        const_add = frame.add(1)
-        tm.assert_frame_equal(const_add, frame + 1)
+        # ndim >= 3
+        ndim_5 = np.ones(float_frame.shape + (3, 4, 5))
```
can you add a comment here on what this is testing

h-vetinari replied:
See above - I did not write the test; just removed indentation

jreback commented:
@h-vetinari ping when updated

h-vetinari committed:
Review (jreback)

h-vetinari commented:
@jreback 

I though your request for `all_arithmetic_operators` would not change anything, but it includes more operators than previously tested - I get errors:
```
AttributeError: module 'operator' has no attribute '__rmod__'  [etc.]
```

I've skipped those tests now.

h-vetinari committed:
PEP8

h-vetinari committed:
Merge remote-tracking branch 'upstream/master' into fixturize_frame_arithmetic

h-vetinari committed:
Split up test_arith_flex_frame (review jreback)

h-vetinari commented:
@ jreback
> might be more clear to move these to a separate test (from here down)

> can you add a comment here on what this is testing

OK, I relented, and split up and cleaned up that huge test. Also the `skips` for the `r`-versions of the ops aren't necessary anymore. They were tested for `add/sub/mult` before, and I extended that to all ops now.

h-vetinari commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-                    tm.assert_frame_equal(result, exp)
-                    _check_mixed_int(result, dtype=dtype)
-
-                    # rops
```
rops were only tested within that if-block. Now they're all tested

h-vetinari commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-                    exp = f(intframe, 2 * intframe)
-                    tm.assert_frame_equal(result, exp)
-
-                    # vs mix int
```
For some reason, this part was tested twice. only once in cleaned-up version

h-vetinari commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+        tm.assert_frame_equal(result, exp)
+        _check_mixed_float(result, dtype=dict(C=None))
+
+    @pytest.mark.parametrize('op', ['__add__', '__sub__', '__mul__'])
```
Broke up the test like you wanted

h-vetinari commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+
+        op = all_arithmetic_operators
+
+        # Check that arrays with dim >= 3 raise
```
> can you add a comment here on what this is testing

Added (and expanded)

h-vetinari commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+            arr = np.ones((1,) * dim)
             msg = "Unable to coerce to Series/DataFrame"
             with tm.assert_raises_regex(ValueError, msg):
-                f(frame, ndim_5)
```
This part is redundant (via the `f` of the `op`) is testing the same thing as getting the `op` from the frame directly

jbrockmendel replied:
Can you separate out these dimension-specific tests?

h-vetinari replied:
Split them out in a separate test

h-vetinari reviewed (commented):
@ jreback
> might be more clear to move these to a separate test (from here down)

> can you add a comment here on what this is testing

OK, I relented, and split up and cleaned up that huge test. Also the `skips` for the `r`-versions of the ops aren't necessary anymore. They were tested for `add/sub/mult` before, and I extended that to all ops now.

h-vetinari commented:
@jreback Green

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-
-            # ndim >= 3
-            ndim_5 = np.ones(frame.shape + (3, 4, 5))
+        op = all_arithmetic_operators  # one instance of parametrized fixture
```
put comments on the prior line

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,92 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+
+        if op.startswith('__r'):
+            # get op without "r" and invert it
+            tmp = getattr(operator, op[:2] + op[3:])
```
make this a named function to avoid the tmp something like

```
def f(x, y):
    # comment
    op = op.replace('__r', '__')
    return getattr(operator, op)

jreback replied:
this comment applies to the prior test actually. I don't think you need this here at all, as op is just add, mul, sub.

did you mean to test the reverse of these as well?

h-vetinari replied:
> make this a named function to avoid the tmp something like [...]

Best I can come up with is the following. Is this what you meant?
```
[...]
# one instance of parametrized fixture
op = all_arithmetic_operators

def f(x, y):
    if op.startswith('__r'):
        # get op without "r" and invert it
        return getattr(operator, op.replace('__r', '__'))(y, x)
    return getattr(operator, op)(x, y)
[...]
```

h-vetinari replied:
> this comment applies to the prior test actually. I don't think you need this here at all, as op is just add, mul, sub. did you mean to test the reverse of these as well?

That was indeed not necessary anymore. Removed

jreback replied:
ok this is fine (your change). note that I suspect we do this in other places, maybe see if you can make a more generalized version of this that we can then include in the pandas.util.testing (but followon-PR)

jreback reviewed (changes requested):
ok test looks a lot nicer. small change, ping on green.

h-vetinari committed:
Review (jreback)

jreback commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,86 @@ def test_df_add_flex_filled_mixed_dtypes(self):
+        def f(x, y):
+            if op.startswith('__r'):
+                # get op without "r" and invert it
+                return getattr(operator, op.replace('__r', '__'))(y, x)
```
@jbrockmendel pls have a look here

jbrockmendel replied:
This is fine.  The alternative would be to do a lookup in core.ops, but I think for a reader who is unfamiliar with core.ops this is clearer.

h-vetinari replied:
Modified comment slightly

jreback commented:
ok by me. @jbrockmendel if you'd have a look and merge if ok.

jbrockmendel commented on pandas/tests/frame/test_arithmetic.py:
```diff
@@ -127,132 +126,86 @@ def test_df_add_flex_filled_mixed_dtypes(self):
-        # tm.assert_frame_equal(res_add, frame + frame)
-        # tm.assert_frame_equal(res_sub, frame - frame)
-        # tm.assert_frame_equal(res_mul, frame * frame)
-        # tm.assert_frame_equal(res_div, frame / (2 * frame))
```
Did this commented-out stuff go somewhere else?  Or just determined to be redundant?

h-vetinari replied:
Determined to be redundant with `test_arith_flex_frame`, which checks all arith_ops, and much more thoroughly.

h-vetinari committed:
Review (jbrockmendel)

jreback commented:
thanks @h-vetinari 

so this is a good example of a change. you moved / parameterized things, but its a single change that is quite straightforward for a reviewer to grok / comment on w/o diving into things too much.
